se3cnn/SO3.py

```diff
@@ -478,32 +476,43 @@ def spherical_harmonics_coeff_to_sphere(coeff, alpha, beta):
     return s
 
 
-
 ################################################################################
 # Linear algebra
 ################################################################################
 
-def get_matrix_kernel(A, eps=1e-10):
-    """
-    Compute an orthonormal basis of the kernel (x_1, x_2, ...)
-    A x_i = 0
-    scalar_product(x_i, x_j) = delta_ij
-
-    :param A: matrix
-    :return: matrix where each row is a basis vector of the kernel of A
-    """
-    _u, s, v = torch.svd(A)
-
-    # A = u @ torch.diag(s) @ v.t()
-    kernel = v.t()[s < eps]
-    return kernel
-
-
-def get_matrices_kernel(As, eps=1e-10):
```

Comment:
I liked having dedicated function to get the null spaces of matrices.
I know I use them only once but they are well defined and we never know...

Reply:
it is not like I much into erasing, but do we really need 
`_u, s, v = torch.svd(A); return v.t()[s < eps]` 
which is no longer referenced anywhere, to be kept in active version of the code, rather than just in history

Reply:
Of course if you don't call it at all, remove it.
My suggestion was to modify it and call it once. But now I understand your concern about the garbage collector and its interaction with function arguments.

Reply:
also, about nullspace, it may be not the best idea to use pytorch svd for that in general, as it targets more largest singular values. 
In contrast, there is `scipy.sparse.linalg.svds` (with 'SM' mode). Despite `sparse` in name it actually accepts dense arrays.    
https://docs.scipy.org/doc/scipy/reference/generated/scipy.sparse.linalg.svds.html#scipy.sparse.linalg.svds

Reply:
Go ahead with any function better than `torch.svd`.
By looking at the `scipy.sparse.linalg.svds` documentation I saw a note:
> This is a naive implementation using ARPACK as an eigensolver on A.H * A or A * A.H, depending on which one is more efficient.

I worrying that making several times `A.T @ A` will add numerical instabilities.

Reply:
why several, just once 

Anyway, as for the task in hand, choice was between `eigh` and `eigsh` due to tighter memory control, and former slightly nicer, because of additional options